WstxValidationException: Unknown reason (at end element </nl:nillableIntElement>) when validating a document with nillable elements

ppalaga opened the pull request:
fix #179
fix #190

ppalaga commented:
@cowtowncoder here is the first sketch of what could become a real fix. Any feedback would be appreciated.

cowtowncoder commented:
With a quick look that makes sense.

My one concern is that the attributes should only be tracked when information is needed (validation enabled, or duplicate check), and avoided otherwise.
This because keeping track with a Maps will add measurable overhead for common case where validation is not enabled.

ppalaga commented:
Thank for the feedback, @cowtowncoder! 

> My one concern is that the attributes should only be tracked when information is needed (validation enabled, or duplicate check), and avoided otherwise. This because keeping track with a Maps will add measurable overhead for common case where validation is not enabled.

+1. If SimpleOutputElement is the right place to implement the attribute tracking, then I can gladly have a look to make the implementation more performant.

ppalaga commented:
> My one concern is that the attributes should only be tracked when information is needed (validation enabled, or duplicate check)

That's actually the case already. All calls of `SimpleOutputElement.checkAttrWrite(String, String, String, String)` are guarded by either `BaseStreamWriter.mCheckAttrs` or by `BaseStreamWriter.mValidator != null`.

ppalaga committed:
Assert specific validation error messages to get more certainty that the
expected error occurred

ppalaga committed:
Extend XMLStreamWriter validation test coverage by checking the writer
validation wherever we test the reader validation

ppalaga committed:
Add a reproducer for #189 'Undefined ID' not thrown when validating with SimpleNsStreamWriter

ppalaga committed:
Add a reproducer for #190 'Element <root> has no attribute "verbose"' not thrown from RepairingNsStreamWriter when validating against a DTD schema

ppalaga committed:
WstxValidationException: Unknown reason (at end element
</nl:nillableIntElement>) when validating a document with nillable
elements fix #179 fix #190

ppalaga commented:
Only the last commit is relevant. The rest is also available in https://github.com/FasterXML/woodstox/pull/191 where we can discuss it separately. 

@cowtowncoder could you please review?

grigoni commented:
hi @cowtowncoder, I'm interested in this fix, can you please have a look? thanks a lot

ppalaga commented:
@grigoni I apologize, this hangs due to legal issues with the Contributor Agreement on the side of my company.

janssendj commented:
Is this still hanging on the CA ?

cowtowncoder commented:
@janssendj No progress that I can see. Although now there are also merge conflicts to resolve.

ppalaga commented:
Yeah, I'd be ready to resolve the merge conflicts once the legal issues are solved.

ppalaga commented:
@gnodet it did not pass the approval procedure on our side.

cowtowncoder commented:
Fixed instead via #245, closing.
